### PR TITLE
fix vite watch mode dep optimizer

### PR DESCRIPTION
**Pull request by patricklx**
- [x] all deps are optimized
- [x] initial dep scan creates all deps
- [x] all optimized chunks are used
- [x] new deps are optimized
- [x] different specifiers use same optimized id
- [ ] addon rebuilds are newly optimized

dep optimizer has multiple phases 
* Scanning, which uses the vite plugins and configured rollup plugins .
* pre bundle, does not use configured rollup plugins.

This is the order 
* inital scan 
* initial scan must provide correct import specifier to vite import analysis plugin, this is probably because we cannot use vite own alias handling 
* pre bundling 
* vite starts and gets requests
* vite checks if some need optimization 
* rules:
1. Specifier must be bare import 
2. Importer cannot be in node modules
3. Needs to be node_resolveable, so follow node resolution without additional plugin resolution 
* if a new dep is detected, vite schedules a new scan

Currently this is broken because of 2 things. app is in rewritten-app. It's actually easy to fix by changing to importer to the original app. Only needed when resolving bare imports.

The more difficult one is rewritten-packages. They must be node resolvable from original app. They are not currently.
I suggest to create a link from `@embroider/rewritten-packages` to the one inside . embroider and rewrite specifiers accordingly.

It currently looks like it's working, because the initial scan detects the deps. But any new dep added later will not be optimized.

fixes:
* rewrite rewritten-app to original app location 
* link rewritten-package to be node resolvable
* esbuild scan needs to pass renamed modules and appjsmatch to vite import analysis 
*

### CODE REVIEW

**Comment by ef4** [comment]
> I suggest to create a link from @embroider/rewritten-packages to the one inside

This is how things used to work. Doing it correctly requires more than just links to the rewritten packages, in general it requires you to rewrite the entire node_modules graph in order to make every package see all the correct dependencies.

And I don't think it's necessary. I think that from the perspective of the rewritten-app, the rewritten-packages don't appear to be inside node_modules. But when we stop rewriting the app, they will.

**Comment by patricklx** [comment]
> > I suggest to create a link from @embroider/rewritten-packages to the one inside
> 
> This is how things used to work. Doing it correctly requires more than just links to the rewritten packages, in general it requires you to rewrite the entire node_modules graph in order to make every package see all the correct dependencies.
> 
> And I don't think it's necessary. I think that from the perspective of the rewritten-app, the rewritten-packages don't appear to be inside node_modules. But when we stop rewriting the app, they will.

The problem is that vite needs a bare import and the importer cannot be in node modules. 

What we are currently for resolving rewritten-packages is doing a rehome inside the rewritten-packages, onto moved-target.js https://github.com/embroider-build/embroider/blob/main/packages/core/src/module-resolver.ts#L954

And here is how vite checks if it should optimise:
Specifier must be bare import
Importer cannot be in node modules
Needs to be node_resolveable, so follow node resolution without additional plugin resolution
https://github.com/vitejs/vite/blob/6c323d5b3ab3cdf81d21bbe965ed3c36aa7f0589/packages/vite/src/node/plugins/resolve.ts#L868



So it's not a problem rewritten-app.

**Comment by patricklx** [comment]
It might be possible to make easier if we could force the addons into the dep optimizer.
Otherwise we really need to exclude them again

File: packages/core/src/module-resolver.ts
Comment (patricklx) [inline packages/core/src/module-resolver.ts:230]: ignored should fallback resolve

File: packages/core/src/module-resolver.ts
Comment (patricklx) [inline packages/core/src/module-resolver.ts:490]: workaround

File: packages/shared-internals/src/package.ts
Comment (patricklx) [inline packages/shared-internals/src/package.ts:29]: workaround for fromFile pointint at /app/tests/package.json

**Review by mansona — CHANGES_REQUESTED** [review]
I just discussed this on the Embroider Triage meeting with @patricklx and I'm going to do my best to summarise the situation here: 

Firstly we started to spec out a way to test if a dependency was optimised correctly in the office hours last week. A good first draft of that work can be seen in https://github.com/embroider-build/embroider/pull/1880 where we run vite in dev mode and try to "audit" the app across HTTP. This will allow us to make sure that the dependencies we want to check are coming from the right place (i.e. from the .vite/deps folder) 

This work is blocked on the refactor that myself and @ef4 are pairing on to allow audits to work across HTTP (and async) which hopefully will land this week.

Once the audit refactor lands we will make sure to provide a suite of tests that show the problems that @patricklx is identifying in this PR (and Discord discussions), we can then check those tests against other work that is going on (e.g. https://github.com/embroider-build/embroider/pull/1779 ) to make sure that they are still failing and we can fix accordingly 👍

**Review by mansona — CHANGES_REQUESTED** [review]
This PR should be split up into individual PRs that address single issues that are verifiable with tests. Especially now that https://github.com/embroider-build/embroider/pull/2029 is merged we should be methodical with how we proceed

**Comment by patricklx** [comment]
fixed in #2135

**Pull request closed without merging**
